packages/edit-site/src/components/page-pages/index.js

```diff
@@ -157,6 +157,21 @@ export default function PagePages() {
 		totalPages,
 	} = useEntityRecords( 'postType', postType, queryArgs );
 
+	// Remove any selected pages that are no longer in the list of visible pages.
+	useEffect( () => {
+		if (
+			selection.some(
+				( id ) => ! pages?.some( ( page ) => page.id === id )
+			)
+		) {
+			setSelection(
+				selection.filter( ( id ) =>
+					pages?.some( ( page ) => page.id === id )
+				)
+			);
+		}
+	}, [ pages, selection ] );
```

Comment:
This can be done automatically before calling the. `onChangeSelection` callback or when calling the `onView` callback with a page change.

Reply:
Hello @youknowriad, I am uncertain about how to approach the issue at hand. It seems that if a page is deleted, neither the onChangeSelection nor the onView functions are triggered. Essentially, we require a mechanism to automatically remove the deleted page (or page that stopped being visible because of a state change) from the selection set if it was previously included.

Reply:
This is another point that I don't get why the selection handling is at the consumer level and not handled internally, but 🤷

Reply:
I don't mind internal selection state but we need an onChangeSelection callback to adapt the UI in site-by-side view.

For the pages removed externally, what's wrong with keeping an outdated selection object with things that are not visible?? Is that bad.

Reply:
I think the selection should definitely be updated `down => top` (dataviews component calling a callback `onSelectionChange` or `onItemsSelected` if you prefer). If a field is updating the outside "selection" object directly, that's a bug (regardless of whether we're maintaining the state internally or externally) for me.

That said, I do see some need of `top => down` selection change potentially in the future (thinking about things like undo/redo...) It's probably too early to think about that though. So whether it's controlled or uncontrolled today doesn't matter much.

Reply:
> If a field is updating the outside "selection" object directly, that's a bug (regardless of whether we're maintaining the state internally or externally) for me.

[This](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/page-pages/index.js#L225) is how it works today: the `title` field sets the selection. 🤔 

So, if I understood correctly, does it sound like we agree to the following plan:

1. selection work the same for preview & bulk actions
2. absorb selection as part of dataview's responsibility, not the consumer's
3. dataviews should expose an `onSelectionChange` callback for the consumer

Given the current implementation for the preview, I'm not sure that we'd be able to make dataviews absorb selection fully in this PR. If @jorgefilipecosta feels confident, go ahead. Otherwise, I guess we could ship this as it is (selection controlled by the consumer) and revisit when something like https://github.com/WordPress/gutenberg/pull/56746 lands.

Does that sound sensible?

Reply:
yes,

I think the issue is that instead of calling the setSelection state, the field should receive a setSelection callback in the render function and call it. Fields could be defined in very different places than where the selection state is hold (either internally or externally)